### PR TITLE
Fix possible deadlock in QueryStatus

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):

- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix possible deadlock on distributed query cancellation.

### CODE REVIEW

File: src/Interpreters/ProcessList.cpp [hunk 1]
```diff
@@ -396,15 +437,18 @@ void QueryStatus::addPipelineExecutor(PipelineExecutor * e)
...
 void QueryStatus::removePipelineExecutor(PipelineExecutor * e)
 {
     std::lock_guard lock(executors_mutex);
-    assert(std::find(executors.begin(), executors.end(), e) != executors.end());
-    std::erase_if(executors, [e](PipelineExecutor * x) { return x == e; });
+    auto it = std::find_if(executors.begin(), executors.end(), [e](const ExecutorHolderPtr & x){ return x->executor == e; });
+    assert(it != executors.end());
+    /// Invalidate executor pointer inside holder.
+    (*it)->remove();
+    executors.erase(it);
```
Comment (Avogar) [inline src/Interpreters/ProcessList.cpp:451]: Actually, here we can just invalidate pointer to executor inside holder and don't erase holder from the executors. In this case we won't need that ugly code in cancelQuery method. And in `~QueryStatus` we can just check that all executors indide holders were invalidated

File: src/Interpreters/ProcessList.cpp [hunk 2]
```diff
@@ -373,16 +373,57 @@ QueryStatus::~QueryStatus()
...
+            /// thread, and query executing thread can call removePipelineExecutor and lock executors_mutex,
+            /// which will lead to deadlock.
+            lock.unlock();
+            e->cancel();
+            lock.lock();
+            cancelled.insert(e.get());
+            /// While executors_mutex was unlocked, removePipelineExecutor could be called and
+            /// the size of executors could have changed. In this case we should start iterating
+            /// over it again to avoid using invalid iterators.
+            if (executors.size() != size)
```
Comment (KochetovNicolai) [inline src/Interpreters/ProcessList.cpp:423]: We can move `executors` to a temporary vector, and then iterate it without holding mutex and iterator invalidation.
We need to do it in a loop as well, cause, probably, after releasing a mutex `executors` could be filled with more items.

File: src/Interpreters/ProcessList.cpp (hunk 1 above)
Comment (KochetovNicolai) [inline src/Interpreters/ProcessList.cpp:451]: I think we can ignore this row now.

Comment (KochetovNicolai) [reply src/Interpreters/ProcessList.cpp:451]: Or maybe use std::list for executors. Looks like we do not need random access.

**Review by KochetovNicolai — COMMENTED** [review]
LGTM

File: src/Interpreters/ProcessList.cpp (hunk 1 above)
Comment (KochetovNicolai) [reply src/Interpreters/ProcessList.cpp:451]: Yes, let's do it.

**Comment by KochetovNicolai** [comment]
Crash is related.
A bit better symbolized trace: https://pastila.nl/?03a8e423/2d9384dc04d6afc1cca2510a7a1a027e

It's in
```
DB::QueryStatus::cancelQuery(bool)

DB::ProcessList::sendCancelToQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool)
/fasttest-workspace/build/../../ClickHouse/src/Interpreters/ProcessList.cpp:534
```

**Comment by Avogar** [comment]
Crash is fixed.

**Comment by KochetovNicolai** [comment]
Looks ok. But, please, fix build :)

**Review by alexey-milovidov — APPROVED** [review]

File: src/Interpreters/ProcessList.cpp [hunk 3]
```diff
@@ -373,16 +377,43 @@ QueryStatus::~QueryStatus()
...
+        /// cancel() can try to lock some internal mutex that is already locked by query executing
+        /// thread, and query executing thread can call removePipelineExecutor and lock executors_mutex,
+        /// which will lead to deadlock.
+        /// Note that the size and the content of executors cannot be changed while
+        /// executors_mutex is unlocked, because:
+        /// 1) We don't allow adding new executors while cancelling query in addPipelineExecutor
+        /// 2) We don't actually remove executor holder from executors in removePipelineExecutor,
+        /// just mark that executor is invalid.
+        /// So, it's safe to continue iteration over executors after subsequent mutex locking.
+        lock.unlock();
```
Comment (novikd) [inline src/Interpreters/ProcessList.cpp:413]: Maybe better not to perform unlock/lock on each iteration, but to collect `executers` snapshot and then for each element call `cancel`?

Comment (Avogar) [reply src/Interpreters/ProcessList.cpp]: Done

**Pull request merged**